packages/langium/src/workspace/index-manager.ts

```diff
@@ -106,38 +108,53 @@ export class DefaultIndexManager implements IndexManager {
         return stream(result);
     }
 
-    allElements(nodeType = ''): Stream<AstNodeDescription> {
-        if (!this.globalScopeCache.has('')) {
-            this.globalScopeCache.set('', Array.from(this.simpleIndex.values()).flat());
+    allElements(nodeType?: string, uris?: Set<string>): Stream<AstNodeDescription> {
+        const allUris = this.documents.all.map(doc => doc.uri.toString());
+        const arrays: AstNodeDescription[][] = [];
+        for (const uri of allUris) {
+            if (!uris || uris.has(uri)) {
+                arrays.push(this.getFileDescriptions(uri, nodeType));
+            }
         }
+        return stream(arrays).flat();
+    }
 
-        const cached = this.globalScopeCache.get(nodeType);
-        if (cached) {
-            return stream(cached);
-        } else {
-            const elements = this.globalScopeCache.get('')!.filter(e => this.astReflection.isSubtype(e.type, nodeType));
-            this.globalScopeCache.set(nodeType, elements);
-            return stream(elements);
+    protected getFileDescriptions(uri: string, nodeType?: string): AstNodeDescription[] {
+        if (!nodeType) {
+            return this.simpleIndex.get(uri) ?? [];
+        }
+        let map = this.simpleTypeIndex.get(uri);
+        if (!map) {
+            map = new Map();
+            this.simpleTypeIndex.set(uri, map);
+        }
+        let descriptions = map.get(nodeType);
+        if (!descriptions) {
+            const allFileDescriptions = this.simpleIndex.get(uri) ?? [];
+            descriptions = allFileDescriptions.filter(e => this.astReflection.isSubtype(e.type, nodeType));
```

Comment:
The result type of `allElements` is a Stream and in most cases the stream is traversed only once, so I think it would make sense to avoid creating intermediate arrays here.
```suggestion
            descriptions = stream(allFileDescriptions).filter(e => this.astReflection.isSubtype(e.type, nodeType));
```

Reply:
> in most cases the stream is traversed only once

That's not correct. The decision to make this an array was made deliberately, as the `descriptions` array is usually only computed once per document change. I.e. if a document changes and requests the global scope during linking, it can reuse the cached `descriptions` arrays of the unaffected documents. Making it into a stream would require to recompute the filter step, which is quite expensive.